tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/union/ControlledWorkerUnion.java

```diff
@@ -61,4 +61,15 @@ public void toss(Exception e) {
 		super.toss(e);
 		phaser.arriveAndDeregister();
 	}
+
+	@Override
+	public void handleClose() throws QueryEvaluationException {
+		try {
+			super.handleClose();
+		} finally {
+			// signal the phaser to close (if currently being blocked)
+			phaser.forceTermination();
+		}
+	}
+
```

Comment:
This was the initial issue. The union() method would deadlock because the phaser was never terminated if the close() method had been called before all the tasks had completed.

There are three other classes that use the same approach with phasers and all those have this exact code.